nemo/collections/asr/modules/audio_modules.py

```diff
@@ -718,6 +718,257 @@ def forward(self, input: torch.Tensor, input_length: torch.Tensor) -> Tuple[torc
+    Args:
+        num_iterations: Number of iterations for the EM algorithm
+        eps: Small value for regularization
+        dtype: Data type for internal computations (default `torch.cdouble`)
+
+    References:
+        [1] Ito et al., Complex Angular Central Gaussian Mixture Model for Directional Statistics in Mask-Based Microphone Array Signal Processing, 2016
+        [2] Boeddeker et al., Front-End Processing for the CHiME-5 Dinner Party Scenario, 2018
+    """
+
+    def __init__(self, num_iterations: int = 3, eps: float = 1e-8, dtype: torch.dtype = torch.cdouble):
+        super().__init__()
+
+        if num_iterations <= 0:
+            raise ValueError(f'Number of iterations must be positive, got {num_iterations}')
+
+        # number of iterations for the EM algorithm
+        self.num_iterations = num_iterations
+
+        if eps <= 0:
+            raise ValueError(f'eps must be positive, got {eps}')
+
+        # small regularization constant
+        self.eps = eps
+
+        # internal calculations
+        if dtype not in [torch.cfloat, torch.cdouble]:
+            raise ValueError(f'Unsupported dtype {dtype}, expecting cfloat or cdouble')
+        self.dtype = dtype
+
+        logging.debug('Initialized %s', self.__class__.__name__)
+        logging.debug('\tnum_iterations: %s', self.num_iterations)
+        logging.debug('\teps:            %g', self.eps)
+        logging.debug('\tdtype:          %s', self.dtype)
+
+    def normalize(self, x: torch.Tensor, dim: int = -3) -> torch.Tensor:
+        """Normalize input to have a unit L2-norm across `dim`.
+        By default, normalizes across the input channels.
+
+        Args:
+            x: C-channel input signal, shape (B, C, F, T)
+            dim: Dimension for normalization, defaults to -3 to normalize over channels
+
+        Returns:
+            Normalized signal, shape (B, C, F, T)
+        """
+        norm_x = torch.linalg.vector_norm(x, ord=2, dim=dim, keepdim=True)
+        x = x / (norm_x + self.eps)
+        return x
+
+    @typecheck(
+        input_types={
+            'alpha': NeuralType(('B', 'C', 'D')),
+            'activity': NeuralType(('B', 'C', 'T')),
+            'log_pdf': NeuralType(('B', 'C', 'D', 'T')),
+        },
+        output_types={'gamma': NeuralType(('B', 'C', 'D', 'T')),},
+    )
+    def update_masks(self, alpha: torch.Tensor, activity: torch.Tensor, log_pdf: torch.Tensor) -> torch.Tensor:
+        """Update masks for the cACGMM.
+
+        Args:
+            alpha: component weights, shape (B, num_outputs, F)
+            activity: temporal activity for the components, shape (B, num_outputs, T)
+            log_pdf: logarithm of the PDF, shape (B, num_outputs, F, T)
+
+        Returns:
+            Masks for the components of the model, shape (B, num_outputs, F, T)
+        """
+        # (B, num_outputs, F)
+        # normalize across outputs in the log domain
+        gamma = log_pdf - torch.max(log_pdf, axis=-3, keepdim=True)[0]
```

Comment:
I think this `gamma` should be either `log_gamma`, or some other variable name since it goes through scale shift through exp function.